### PR TITLE
En 8042 start in epoch in import db

### PR DESCRIPTION
**Pull request by iulianpascalau**
- implemented trie storage resolver, added some import-db fixes, prepared the ground for the start in epoch in import-db process

### CODE REVIEW

File: dataRetriever/storageResolvers/headerResolver.go [hunk 1]
```diff
@@ -106,7 +107,29 @@ func (hdrRes *headerResolver) RequestDataFromNonce(nonce uint64, epoch uint32) e
 
 // RequestDataFromEpoch requests the epoch start block
 func (hdrRes *headerResolver) RequestDataFromEpoch(identifier []byte) error {
-	return hdrRes.RequestDataFromHash(identifier, 0)
+	buff, err := hdrRes.resolveHeaderFromEpoch(identifier)
+	if err != nil {
+		hdrRes.signalGracefullyClose()
+
```
Comment (sasurobert) [inline dataRetriever/storageResolvers/headerResolver.go:113]: delete empty line

File: dataRetriever/storageResolvers/trieNodeResolver.go [hunk 2]
```diff
@@ -1,26 +1,135 @@
...
+	}
+
+	return tnr.sendDataToSelf(nodes)
+}
+
+func (tnr *trieNodeResolver) getSubTrie(hash []byte, remainingSpace uint64) ([][]byte, uint64, error) {
+	serializedNodes, remainingSpace, err := tnr.trieDataGetter.GetSerializedNodes(hash, remainingSpace)
+	if err != nil {
+		tnr.signalGracefullyClose()
+
```
Comment (sasurobert) [inline dataRetriever/storageResolvers/trieNodeResolver.go:108]: delete empty line

File: integrationTests/factory/genesis.json [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    "address": "erd1yajssshtsc75x87cxvylnwu4r9dv3c2tegufrd07fjmw72krlq9spmw32d",
+    "supply": "2222222222222222222222222",
+    "balance": "2219722222222222222222222",
+    "stakingvalue": "0",
+    "delegation": {
+      "address": "erd1qqqqqqqqqqqqqpgqrchxzx5uu8sv3ceg8nx8cxc0gesezure5awqn46gtd",
+      "value": "2500000000000000000000"
+    }
+  }
+]
```
Comment (sasurobert) [inline integrationTests/factory/genesis.json:92]: empty line at end of file

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: dataRetriever/storageResolvers/headerResolver.go (hunk 1 above)
Comment (iulianpascalau) [reply dataRetriever/storageResolvers/headerResolver.go:113]: done

File: dataRetriever/storageResolvers/trieNodeResolver.go (hunk 2 above)
Comment (iulianpascalau) [reply dataRetriever/storageResolvers/trieNodeResolver.go:108]: done

File: integrationTests/factory/genesis.json (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/factory/genesis.json:92]: done

File: integrationTests/factory/consensusComponents/consensusComponents_test.go [hunk 4]
```diff
@@ -30,7 +30,13 @@ func TestConsensusComponents_Close_ShouldWork(t *testing.T) {
 	require.Nil(t, err)
 	managedNetworkComponents, err := node.CreateManagedNetworkComponents(configs, managedCoreComponents)
 	require.Nil(t, err)
-	managedBootstrapComponents, err := node.CreateManagedBootstrapComponents(configs, managedCoreComponents, managedCryptoComponents, managedNetworkComponents)
+	managedBootstrapComponents, err := node.CreateManagedBootstrapComponents(
+		configs,
+		managedCoreComponents,
+		managedCryptoComponents,
+		managedNetworkComponents,
+		make(chan endProcess.ArgEndProcess),
```
Comment (raduchis) [inline integrationTests/factory/consensusComponents/consensusComponents_test.go:38]: I think you can reuse the chan from the managedCoreComponents, it should be the same, right? There should be no extra parameter in this method

File: node/nodeRunner.go [hunk 5]
```diff
@@ -778,7 +785,7 @@ func CreateManagedDataComponents(configs *config.Configs, managedCoreComponents
 		Core:                          managedCoreComponents,
 		EpochStartNotifier:            managedCoreComponents.EpochStartNotifierWithConfirm(),
 		CurrentEpoch:                  storerEpoch,
-		CreateTrieEpochRootHashStorer: configs.FlagsConfig.ImportDbSaveTrieEpochRootHash,
+		CreateTrieEpochRootHashStorer: configs.ImportDbConfig.ImportDbSaveTrieEpochRootHash,
```
Comment (raduchis) [inline node/nodeRunner.go:788]: there is no check for importDBEnabled here. Should the SaveTrieEpochRootHash be enabled also when not running in importDb mode? If so, it should be move from the ImportDBConfig to a more general config location.

**Review by raduchis — CHANGES_REQUESTED** [review]

File: cmd/node/flags.go [hunk 6]
```diff
@@ -374,21 +382,28 @@ func applyFlags(ctx *cli.Context, cfgs *config.Configs, log logger.Logger) {
...
-	flagsConfig.IsInImportMode = isInImportMode
-	flagsConfig.ImportDbNoSigCheckFlag = importDbNoSigCheckFlag
-	flagsConfig.ImportDbSaveTrieEpochRootHash = ctx.GlobalBool(importDbSaveEpochRootHash.Name) && isInImportMode
-
+	importDBConfigs := &config.ImportDbConfig{
+		IsImportDBMode:                len(importDbDirectoryValue) > 0,
+		ImportDBWorkingDir:            importDbDirectoryValue,
+		ImportDbNoSigCheckFlag:        ctx.GlobalBool(importDbNoSigCheck.Name),
+		ImportDbSaveTrieEpochRootHash: ctx.GlobalBool(importDbSaveEpochRootHash.Name),
+		ImportDBStartInEpoch:          uint32(ctx.GlobalUint64(importDbStartInEpoch.Name)),
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/flags.go:390]: Will be used when the feature is fully functional. The existing `start-in-epoch` is of type bool. That flag will be phased out when we will not support start in epoch = false "feature"

File: cmd/storer2elastic/dataprocessor/ratingsProcessor.go [hunk 7]
```diff
@@ -120,18 +119,7 @@ func (rp *ratingsProcessor) indexRating(epoch uint32, validatorsRatingData map[u
...
-		fmt.Sprintf("%s_%s", "Shard", core.PathShardPlaceholder),
-		core.PathIdentifierPlaceholder)
-
-	pathTemplateForStaticStorer := filepath.Join(
-		rp.dbPathWithChainID,
-		core.DefaultStaticDbString,
-		fmt.Sprintf("%s_%s", "Shard", core.PathShardPlaceholder),
-		core.PathIdentifierPlaceholder)
-	pathManager, err := pathmanager.NewPathManager(pathTemplateForPruningStorer, pathTemplateForStaticStorer)
+	pathManager, err := storageFactory.CreatePathManager(rp.dbPathWithChainID, rp.dbPathWithChainID)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/storer2elastic/dataprocessor/ratingsProcessor.go:122]: I felt into my own trap. I usually state in PRs that it is a bad practice to have consecutive parameters of the same type as they can easily be switched between. Good find!

File: cmd/node/flags.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/flags.go:390]: fixed

File: cmd/storer2elastic/dataprocessor/ratingsProcessor.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/storer2elastic/dataprocessor/ratingsProcessor.go:122]: fixed

File: integrationTests/factory/consensusComponents/consensusComponents_test.go (hunk 4 above)
Comment (iulianpascalau) [reply integrationTests/factory/consensusComponents/consensusComponents_test.go:38]: right, we can reuse that one :+1:

Comment (iulianpascalau) [reply integrationTests/factory/consensusComponents/consensusComponents_test.go:38]: done

File: node/nodeRunner.go (hunk 5 above)
Comment (iulianpascalau) [reply node/nodeRunner.go:788]: was not decided if we need this besides the scope of the import-DB process.
Will keep it close to the import-db process, will make all flags from import-db config structure become false if the isInImportDBMode is false.

Comment (iulianpascalau) [reply node/nodeRunner.go:788]: done

File: cmd/node/flags.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/flags.go:390]: ok!

File: cmd/storer2elastic/dataprocessor/ratingsProcessor.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/storer2elastic/dataprocessor/ratingsProcessor.go]: ok

File: integrationTests/factory/consensusComponents/consensusComponents_test.go (hunk 4 above)
Comment (raduchis) [reply integrationTests/factory/consensusComponents/consensusComponents_test.go]: ok!

**Review by raduchis — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**